core/client/fs/src/main/java/alluxio/client/file/cache/filter/ConcurrentClockCuckooFilter.java

```diff
@@ -0,0 +1,1070 @@
+    for (int i = 0; i < maxPathLen; i++) {
+      cuckooPath[i] = new CuckooRecord();
+    }
+    boolean done = false;
+    while (!done) {
+      int depth = cuckooPathSearch(b1, b2, fp, cuckooPath);
+      if (depth < 0) {
+        break;
+      }
+      if (cuckooPathMove(b1, b2, fp, cuckooPath, depth)) {
+        pos.setBucketAndSlot(cuckooPath[0].mBucket, cuckooPath[0].mSlot);
+        pos.setStatus(CuckooStatus.OK);
+        done = true;
+      }
+    }
+    if (!done) {
+      // NOTE: since we assume holding the locks of two buckets before calling this method,
+      // we keep this assumptions after return.
+      mLocks.lockTwoWrite(b1, b2);
+    }
+    return done;
+  }
+
+  /**
+   * Search for an empty slot from two initial buckets b1 and b2, and move items backwards along the
+   * path.
+   *
+   * @param b1 the first bucket
+   * @param b2 the second bucket
+   * @param fp the fingerprint
+   * @param cuckooPath the information of searched path
+   * @return the depth the path
+   */
+  private int cuckooPathSearch(int b1, int b2, int fp, CuckooRecord[] cuckooPath) {
+    // 1. search a path
+    BFSEntry x = slotBFSSearch(b1, b2, fp);
+    if (x.mDepth == -1) {
+      return -1;
+    }
+    // 2. re-construct path from x
+    for (int i = x.mDepth; i >= 0; i--) {
+      cuckooPath[i].mSlot = x.mPathcode % TAGS_PER_BUCKET;
+      x.mPathcode /= TAGS_PER_BUCKET;
+    }
+    if (x.mPathcode == 0) {
+      cuckooPath[0].mBucket = b1;
+    } else {
+      assert x.mPathcode == 1;
+      cuckooPath[0].mBucket = b2;
+    }
+    {
+      mLocks.lockOneWrite(cuckooPath[0].mBucket);
+      int tag = mTable.readTag(cuckooPath[0].mBucket, cuckooPath[0].mSlot);
+      if (tag == 0) {
+        mLocks.unlockOneWrite(cuckooPath[0].mBucket);
+        return 0;
+      }
+      mLocks.unlockOneWrite(cuckooPath[0].mBucket);
+      cuckooPath[0].mFingerprint = tag;
+    }
+    for (int i = 1; i <= x.mDepth; i++) {
+      CuckooRecord curr = cuckooPath[i];
+      CuckooRecord prev = cuckooPath[i - 1];
+      curr.mBucket = altIndex(prev.mBucket, prev.mFingerprint);
+      mLocks.lockOneWrite(curr.mBucket);
+      int tag = mTable.readTag(curr.mBucket, curr.mSlot);
+      if (tag == 0) {
+        mLocks.unlockOneWrite(curr.mBucket);
+        return i;
+      }
+      curr.mFingerprint = tag;
+      mLocks.unlockOneWrite(curr.mBucket);
+    }
+    return x.mDepth;
+  }
+
+  /**
+   * Search an empty slot from two initial buckets.
+   *
+   * @param b1 the first bucket
+   * @param b2 the second bucket
+   * @param fp the fingerprint
+   * @return the last entry of searched path
+   */
+  private BFSEntry slotBFSSearch(int b1, int b2, int fp) {
+    Queue<BFSEntry> queue = new LinkedList<>();
+    queue.offer(new BFSEntry(b1, 0, 0));
+    queue.offer(new BFSEntry(b2, 1, 0));
+    int maxPathLen = MAX_BFS_PATH_LEN;
+    while (!queue.isEmpty()) {
+      BFSEntry x = queue.poll();
+      mLocks.lockOneWrite(x.mBucket);
```

Comment:
It might be better to use a closable lockResource, then we don't have to release the lock explicitly. 
```
 try (LockResource r = new LockResource(lock)) {
      ...
 }
```
see alluxio.resource.LockResource